### PR TITLE
feat: comprehensive bucket validation for s3, gcs, and azure

### PR DESCRIPTION
**Pull request by bwalsh**
## Summary

This PR introduces comprehensive bucket name validation across storage providers (S3, GCS, Azure), migrates validation logic to a centralized location, and updates the core API to enforce these rules. It also expands test coverage to include invalid edge cases and provider-specific naming conventions.

## Why

Bucket naming rules vary between cloud providers. Previously, Syfon only performed minimal validation, leading to runtime failures during URL signing or storage operations. Centralizing and enforcing these rules at the config and API levels ensures "fail-fast" behavior and consistent error reporting.

### Naming & Validation (`internal/common/cloud.go`)
- Migrated naming regexes and validation logic from `internal/provider` (deprecated).
- Implemented `ValidateBucketName` which applies provider-specific rules:
  - **S3**: 3-63 characters, lowercase letters, numbers, and hyphens only; no dots for S3-managed buckets.
  - **GCS**: Supports dotted names (up to 222 chars), underscores, and length up to 63 for simple names; rejects IP addresses and "goog" prefixes.
  - **Azure**: 3-63 characters, lowercase letters, numbers, and hyphens; no consecutive hyphens.

### Configuration (`internal/config/config.go`)
- Enforces `ValidateBucketName` during `LoadConfig` for all static credentials.
- Normalizes provider aliases (e.g., `gs` -> `gcs`, `azblob` -> `azure`).

### API Handlers (`internal/api/internaldrs/buckets.go`)
- Updated `handleInternalPutBucketFiber` and `handleInternalCreateBucketScopeFiber` to validate bucket names on dynamic creation.
- Reconciled merge conflicts with Fiber v3 architecture.

### Tests
- **`internal/config/config_test.go`**: Expanded to test:
  - Valid bucket names (S3 format).
  - Invalid bucket names (length, dots, special characters).
  - Non-S3 provider specific validation (GCS dotted names, Azure hyphens).
  - Provider normalization logic.
- **`internal/common/cloud_test.go`**: New unit tests for provider-specific regex logic.
- **`internal/api/internaldrs/data_test.go`**: Updated integration tests to use valid bucket names for Gen3-authorized operations.

## Testing

Verified with full test suite:
```bash
go test ./...
```

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -306,6 +309,12 @@ func LoadConfig(configFile string) (*Config, error) {
 		if cred.Bucket == "" {
 			return nil, fmt.Errorf("s3_credentials[%d]: bucket is required", i)
 		}
+		if len(cred.Bucket) < 3 || len(cred.Bucket) > 63 {
+			return nil, fmt.Errorf("s3_credentials[%d]: bucket name %q must be 3–63 characters", i, cred.Bucket)
+		}
+		if !validBucketName.MatchString(cred.Bucket) {
+			return nil, fmt.Errorf("s3_credentials[%d]: bucket name %q is invalid (lowercase letters, numbers, hyphens only; must start and end with letter or number)", i, cred.Bucket)
```
[the remark this answers is not in the document]
Comment (bwalsh) [reply config/config.go]: done

File: config/config.go [hunk 2]
```diff
@@ -296,6 +297,8 @@ func LoadConfig(configFile string) (*Config, error) {
 		return nil, fmt.Errorf("no database specified in config")
 	}
 
+	var validBucketName = regexp.MustCompile(`^[a-z0-9][a-z0-9\-]{1,61}[a-z0-9]$`)
+
```
[the remark this answers is not in the document]
Comment (bwalsh) [reply config/config.go]: done

**Review by matthewpeterkort — CHANGES_REQUESTED** [review]
### PR Feedback: Expand validation to support non-S3 providers
Since this PR was opened, support for non-S3 providers (`gcs`, `azure`, and `file`) has been integrated into the `s3_credentials` configuration. Stricter validation is a great addition, but it needs to account for the different naming rules of these providers so we don't accidentally reject valid local paths or GCS buckets.
#### Suggested Changes
**1. Create a provider-aware validator**
Instead of a global S3-only validator, let's move the logic to a helper that checks the `provider` type. A good location for this is `syfon/internal/provider/validation.go`:
```go
package provider
import (
	"fmt"
	"regexp"
	"strings"
)
var (
	// S3/GCS: 3-63 chars, lowercase, numbers, hyphens, dots.
	reS3 = regexp.MustCompile(`^[a-z0-9][a-z0-9.-]{1,61}[a-z0-9]$`)
	// Azure: 3-63 chars, lowercase, numbers, hyphens.
	reAzure = regexp.MustCompile(`^[a-z0-9](-?[a-z0-9]){2,62}$`)
)
func ValidateBucketName(bucket string, p string) error {
	p = Normalize(p, S3)
	bucket = strings.TrimSpace(bucket)
	if bucket == "" {
		return fmt.Errorf("bucket name cannot be empty")
	}
	switch p {
	case S3, GCS:
		if !reS3.MatchString(bucket) {
			return fmt.Errorf("invalid %s bucket name %q: must be 3-63 characters (lowercase, numbers, dots, or hyphens)", p, bucket)
		}
	case Azure:
		if !reAzure.MatchString(bucket) {
			return fmt.Errorf("invalid azure container name %q: must be 3-63 characters (lowercase, numbers, or hyphens)", bucket)
		}
	case File:
		// For local filesystem, we just ensure it's not empty. 
		// Local paths like "/tmp/data" are valid here but would fail S3 regex.
		return nil
	}
	return nil
}
```

**2. Update the config loading loop In syfon/config/config.go, update the validation loop to pass the current provider to the validator**

```for i, cred := range cfg.S3Credentials {
    // ... existing provider normalization ...
    if err := provider.ValidateBucketName(cred.Bucket, cred.Provider); err != nil {
        return nil, fmt.Errorf("s3_credentials[%d]: %w", i, err)
    }
    // ...
}
```

***3. Test Coverage Please update the tests to include:***

* A valid local path under provider: file (e.g., /mnt/data/my-bucket).
* A valid GCS/S3 name (e.g., my-bucket-name).
* An invalid Azure name (e.g., my_bucket - underscores are forbidden in Azure).

**Review by matthewpeterkort — APPROVED** [review]
